### PR TITLE
[Feature] try-runtime bring-up

### PR DESCRIPTION
**Pull request by shamb0**
# Pull Request type

Enabling `try-runtime` feature of substrate

Please add the labels corresponding to the type of changes your PR introduces:

- Feature

## What is the current behavior?

* The code has been refactored to integrate the `try-runtime` features of Substrate. The execution of `try-runtime` sub-commands has been successfully performed and the results are captured below.

* For detailed HOWTO's and the test report, please refer to [docs/10-01-howto-feature-try-runtime.md].

**Sub-Command Execution Status**

| Test ID | Sub-Command | Status |
| --- | --- | --- |
| 1 | create-snapshot | OK |
| 2 | on-runtime-upgrade (live) | OK |
| 3 | on-runtime-upgrade (snap) | OK |
| 4 | execute-block (live) | OK |
| 5 | execute-block (snap) | OK |
| 6 | offchain-worker (live) | OK |
| 7 | offchain-worker (snap) | OK |
| 8 | follow-chain | OK |
| 9 | fast-forward (live) | KO |
| 10 | fast-forward (snap) | KO |

Tests with ID 9 and 10 have encountered failures, with the error message "ERROR main runtime: panicked at 'Timestamp slot must match `CurrentSlot`'". The underlying cause of this issue is currently unknown. To facilitate the resolution and ensure progress, a support ticket has been raised and submitted to the substrate. For more information, please visit [here](https://substrate.stackexchange.com/questions/9024/substrate-try-runtime-sub-command-fast-forward-error-main-runtime-pa).

### CODE REVIEW

**Comment by shamb0** [comment]
Hi @abdelhamidbakhta, The pull request is ready for review. I would appreciate it if you could take a look at it. Should we discuss the possibility of disregarding the codecov check?

File: docs/10-01-howto-feature-try-runtime.md [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+local testnet.
+
+================================================================
+
+## create-snapshot
+
+> Step-01 | create snapshot from sharingan testnet
+
+```bash
+export RPC_WS_URL="ws://127.0.0.1:9931" && \
```
Comment (EvolveArt) [inline docs/10-01-howto-feature-try-runtime.md:63]: 9931 ?

File: crates/runtime/src/lib.rs [hunk 2]
```diff
@@ -356,22 +356,23 @@ impl_runtime_apis! {
 
     #[cfg(feature = "try-runtime")]
     impl frame_try_runtime::TryRuntime<Block> for Runtime {
-        fn on_runtime_upgrade() -> (Weight, Weight) {
+        fn on_runtime_upgrade(checks: frame_try_runtime::UpgradeCheckSelect) -> (Weight, Weight) {
```
Comment (EvolveArt) [inline crates/runtime/src/lib.rs:359]: explain the rationale behind this new argument in comments pls

**Review by EvolveArt — CHANGES_REQUESTED** [review]
Thanks a lot for this, would you mind presenting a demo at the next Community Call on Friday ? Would love to see done in practice against sharingan testnet `sharingan.madara.wtf` RPC.

Just a few questions :
- Why define a new chain spec ? cfg2 ?
- Why adding a signature check argument ? I doubt this would be useful in our case ?

This does pave the way for runtime upgrades which we'd still have to define the process. Who is able to trigger them ? How could we link that to an on-chain multisig for instance ?

**Comment by shamb0** [comment]
> 

Hi @EvolveArt,

Thank you for your review feedback!

- I'm available to conduct the demo. Can we schedule it for the Community Call on July 7th? Does that work for you?

- The patch has not yet been tested on the Sharingan Testnet. Last week, the testnet was temporarily unavailable due to ongoing migration work. As an alternative, @abdelhamidbakhta suggested conducting the test on the local testnet.

> **Why define a new chain spec? cfg2?**

- That's an excellent question! I noticed that the testnet launch script `infra/local-testnet/run.sh` doesn't function properly when using the `--chain=local` chainspec. Initial investigations revealed an issue with the bootnode address. Additionally, when using `--chain=local`, six PoA addresses are initialized as part of the chainspec genesis. Therefore, we need at least five validator nodes to finalize the block.

- To address these challenges, I made the following changes in the new testnet launch script `infra/local-testnet/run-cfg2.sh`:

      - Created a new chainspec called `crates/node/chain-specs/madara-local-cfg2.json` based on `--chain=local`, disregarding the default configuration of `"bootNodes": []`.

      - Defined the bootnode as "boot-node-a" with the corresponding node key "bb2f8c969875fe4bd5ad7275b92e0b80a9f424243fe237fbd992d5990197e01d".

      - Established the authority node as "validator-node-a-to-e" and the full node as "full-node-a". These nodes utilize the public node key of "boot-node-a" for the bootnodes configuration: `--bootnodes=/ip4/127.0.0.1/tcp/30331/p2p/12D3KooWMpdTSKB2FiCRCZsCA2JfASAZUQRFYkvrcQgq5mtEt1fw"`

These are the highlights of the changes. Please let me know if anything is unclear.

> **Why add a signature check argument? I doubt this would be useful in our case?**

Could you please provide additional details about the question? Are you referring to the trait implementation `impl frame_try_runtime::TryRuntime<Block> for Runtime`?

> **This does pave the way for runtime upgrades which we'd still have to define the process. Who is able to trigger them? How could we link that to an on-chain multisig, for instance?**

- You're absolutely right. The process of runtime storage migration on a live chain is different. The documentation provided below explains the process using pallet-utility and sudo for authorization. I can explore the possibility of using on-chain multi-sig for authorization and provide an update on the upgrade procedure.

  - [Substrate Runtime Upgrade, Storage Migration and Rehearsal - HackMD](https://hackmd.io/lf9dq2b3R_SzCA7TscmdgQ?view)

  - [Upgrade a running network | Substrate_ Docs](https://docs.substrate.io/tutorials/build-a-blockchain/upgrade-a-running-network/)

- If time permits during the Community Call, we can also demonstrate the runtime storage migration on the local testnet.

Please let me know if you have any further questions or if there's anything else I can assist you with.

File: docs/10-01-howto-feature-try-runtime.md (hunk 1 above)
Comment (shamb0) [reply docs/10-01-howto-feature-try-runtime.md:63]: Here is the setup for the local testnet and port mapping, where `9931` is used as the RPC endpoint from `boot-node-a`.

| node            | p2p port | RPC port | prometheus port |
| --------------- | -------- | -------- | --------------- |
| boot-node-a     | 30331    | 9931     | 9611            |
| validator-node-a| 30332    | 9932     | 9621            |
| validator-node-b| 30333    | 9933     | 9631            |
| validator-node-c| 30334    | 9934     | 9641            |
| validator-node-d| 30335    | 9935     | 9651            |
| validator-node-e| 30336    | 9936     | 9661            |
| full-node-a     | 30344    | 9944     | -               |

File: crates/runtime/src/lib.rs (hunk 2 above)
Comment (shamb0) [reply crates/runtime/src/lib.rs:359]: These requirements are sourced from Substrate ver "polkadot-v0.9.43", and the code has been refactored to accommodate the following interface changes:

- The internal function signature of the `TryRuntime` trait has been modified. You can find the updated code at the following link:

    [Substrate - TryRuntime](https://github.com/paritytech/substrate/blob/polkadot-v0.9.43/frame/try-runtime/src/lib.rs)

    ```
    fn on_runtime_upgrade(checks: UpgradeCheckSelect) -> (Weight, Weight);
    ```

- The interface definition of "frame-executive" has also been updated. You can refer to the updated code at the following link:

    [Substrate - frame-executive](https://github.com/paritytech/substrate/blob/polkadot-v0.9.43/frame/executive/src/lib.rs)

    ```
    pub fn try_runtime_upgrade(
        checks: frame_try_runtime::UpgradeCheckSelect,
    ) -> Result<Weight, &'static str> {

    }
    ```

Comment (LucasLvy) [reply crates/runtime/src/lib.rs:359]: I think he meant in the code not in a github comment

Comment (shamb0) [reply crates/runtime/src/lib.rs:365]: Whoops, sorry about my bad interpretation. I've added the comment to the function description, thanks for pointing it out!

File: docs/10-01-howto-feature-try-runtime.md (hunk 1 above)
Comment (d-roak) [reply docs/10-01-howto-feature-try-runtime.md:63]: > Why define a new chain spec ? cfg2 ?
We shouldn't do this. @shamb0 please add the generation of the spec on the script to a tmp folder and use it directly from there. We will drop the chainspecs on our repo in the near future (they'll be retrieved from an URL)

**Comment by d-roak** [comment]
> Hi @abdelhamidbakhta @EvolveArt, The pull request is ready for review. I would appreciate it if you could take a look at it. Should we discuss the possibility of disregarding the codecov check?

We can disregard the codecov (if it's a low coverage diff). The remaining steps on CI should be solved

File: docs/10-01-howto-feature-try-runtime.md (hunk 1 above)
Comment (shamb0) [reply docs/10-01-howto-feature-try-runtime.md:63]: Sure, @d-roak, that sounds good. I will make the modifications as you suggested.

**Comment by shamb0** [comment]
Hi @d-roak,

It appears that the latest changes in `crates/node/src/chain_spec.rs` [upstream/main] are causing the following error.  Could you please investigate this issue on your end? I'm curious to know what might have caused it.



**Commands Used**

> **boot-node**

```
./target/release/madara \
	--chain=local 	\
	--alice \
	--validator \
	--force-authoring 	\
	--rpc-cors=all 	\
	--rpc-external 	\
	--rpc-methods=unsafe \
	--node-key bb2f8c969875fe4bd5ad7275b92e0b80a9f424243fe237fbd992d5990197e01d \
	--port 30331 \
	--rpc-port 9931 \
	--base-path /tmp/madara/boot-node-a \
	--madara-path /tmp/madara/boot-node-a-madara
```

> **bob**

```
./target/release/madara \
	--chain=local 	\
	--bob \
	--validator \
	--force-authoring \
	--rpc-cors=all \
	--rpc-external 	\
	--rpc-methods=unsafe \
	--node-key f6d57e8a12ea5467a0d0f632bf3dce14c199c217833fe944ddb912e397b3809c \
	--bootnodes=/ip4/127.0.0.1/tcp/30331/p2p/12D3KooWMpdTSKB2FiCRCZsCA2JfASAZUQRFYkvrcQgq5mtEt1fw \
	--port 30332 \
	--rpc-port 9932 \
	--prometheus-port 9621 \
	--base-path /tmp/madara/validator-node-a \
	--madara-path /tmp/madara/validator-node-a-madara
```

**Comment by d-roak** [comment]
> It appears that the latest changes in `crates/node/src/chain_spec.rs` [upstream/main] are causing the following error. Could you please investigate this issue on your end? I'm curious to know what might have caused it.

Can u tell me a previous commit where those commands were working? You are not booting from a chain spec, so the genesis for the two nodes will be different all the times

**Comment by shamb0** [comment]
> > It appears that the latest changes in `crates/node/src/chain_spec.rs` [upstream/main] are causing the following error. Could you please investigate this issue on your end? I'm curious to know what might have caused it.
> 
> Can u tell me a previous commit where those commands were working? You are not booting from a chain spec, so the genesis for the two nodes will be different all the times

With the last commit ID "a605da56bb81254016432f90f970bcd35f27a0de" (Commits on Jun 26, 2023), everything is working fine.

I used the script from [this link](https://github.com/shamb0/madara/blob/shamb0/dev-01-chainspec-regression-test/infra/local-testnet/run-cfg3.sh) to launch the local testnet. I didn't observe any sync errors, and the block finalization looks okay.

Screenshot

It seems that the commit ID "57d0a3abd5524d7e57c0e88021c3425446552cb8" (Commits on Jun 27, 2023) is causing the regression.

**Comment by abdelhamidbakhta** [comment]
closing as stale for now. please reopen when you actively work on it @shamb0

**Pull request closed without merging**